### DISCUSSION TITLE
http client implementation

### DISCUSSION THREAD

**Discussion by seife** [schreibfaul1/ESP32-audioI2S#264] [Q&A]
Hi,
while looking around the Audio.cpp file I was wondering if it would be beneficial to use the HTTPClient for request handling instead of implementing this all on our own.
Before I start trying this out, I'd like to shorly discuss if the current implementation "just happened" and nobody thought about using HTTPClient -- then I would go ahead and try it -- or if actually HTTPClient was tried before but abandoned due to technical reasons.

From my (maybe naive...) look at it, it seems using HTTPClient would simplify many things (http/https handled transparently, HTTP return code parsing etc...)

Any information is appreciated :-)

**Comment by FrankBoesing** [comment]
Yes, thought this too,. 
But httpclient has some problems with streams. There is no clean way to stop a stream (often, it does not work to stop)
It tries to clear the buffer first - but during that, it continues receiving, so this fails often. Result are weird errormessages or even crashes.
But maybe I've just used it wrong? I'm still pretty new to ESP..

**Reply by seife** [reply]
Ok, good to know this has already been tried.
Then I'll prioritize other things first ;-)

**Reply by FrankBoesing** [reply]
> Then I'll prioritize other things first ;-)

A few simple edits: There are long chains of if...if...if... I guess replacing them with "if..else if..else if" could be beneficial, esp. in the audio loop

**Reply by schreibfaul1** [reply]
Anything that contributes to improvements or simplifications is welcome. I can't contribute anything to the http client, I've never used it.

**Reply by CelliesProjects** [reply]
@FrankBoesing @seife Take a look at https://github.com/CelliesProjects/ESP32_VS1053_Stream/blob/master/src/ESP32_VS1053_Stream.cpp#L383-L405
The stream can be stopped clean with 
```c++
    if (_http) {
        if (_http->connected()) {
            WiFiClient* const stream = _http->getStreamPtr();
            stream->stop();
            stream->flush(); // it seems that since 2.0.0 stream->flush() is not needed anymore after a stream->close();
        }
        _http->end();
        delete _http;
        _http = NULL; 
```
Not sure (but pretty positive it will) if it will work in this lib though...

See the example at https://github.com/espressif/arduino-esp32/blob/master/libraries/HTTPClient/examples/BasicHttpsClient/BasicHttpsClient.ino

and https://github.com/espressif/arduino-esp32/blob/master/libraries/HTTPClient/examples/StreamHttpClient/StreamHttpClient.ino

**Reply by seife** [reply]
> > Then I'll prioritize other things first ;-)
> 
> A few simple edits: There are long chains of if...if...if... I guess replacing them with "if..else if..else if" could be beneficial, esp. in the audio loop

Even though they look like cheap wins, the compiler seems to optimize that already quite well: I could not see that the generated code was really different.
I also tried to change some of them to switch...case constructs, but that actually increased the code size ;-)

**Reply by FrankBoesing** [reply]
> > > Then I'll prioritize other things first ;-)
> > 
> > 
> > A few simple edits: There are long chains of if...if...if... I guess replacing them with "if..else if..else if" could be beneficial, esp. in the audio loop
> 
> Even though they look like cheap wins, the compiler seems to optimize that already quite well: I could not see that the generated code was really different. I also tried to change some of them to switch...case constructs, but that actually increased the code size ;-)

interesting.. :)

**Comment by FrankBoesing** [comment]
Yes...it calls http->stop.... with does not work for me with a stream for the reason i mentioned....
